test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -1421,3 +1513,136 @@ UCS_TEST_P(test_ucp_sockaddr_protocols, am_zcopy_64k,
     UCP_INSTANTIATE_TEST_CASE_TLS(_test_case, all,   "all")
 
 UCP_INSTANTIATE_CM_TEST_CASE(test_ucp_sockaddr_protocols)
+
+
+class test_ucp_sockaddr_protocols_err : public test_ucp_sockaddr_protocols {
+public:
+    static void get_test_variants(std::vector<ucp_test_variant>& variants) {
+        uint64_t features = UCP_FEATURE_TAG;
+        test_ucp_sockaddr::get_test_variants_mt(variants, features,
+                                                SEND_STOP | TEST_MODIFIER_CM,
+                                                "send_stop");
+        test_ucp_sockaddr::get_test_variants_mt(variants, features,
+                                                RECV_STOP | TEST_MODIFIER_CM,
+                                                "recv_stop");
+        test_ucp_sockaddr::get_test_variants_mt(variants, features,
+                                                SEND_STOP | RECV_STOP |
+                                                TEST_MODIFIER_CM,
+                                                "bidi_stop");
+    }
+
+protected:
+    static void recv_stop(void *arg) {
+        test_ucp_sockaddr_protocols_err *test =
+            static_cast<test_ucp_sockaddr_protocols_err*>(arg);
+        test->one_sided_disconnect(test->receiver(), UCP_EP_CLOSE_MODE_FORCE);
+        while (m_err_count == 0) {
+            test->short_progress_loop();
+        }
+    }
+
+    static void send_stop(void *arg) {
+        test_ucp_sockaddr_protocols_err *test =
+            static_cast<test_ucp_sockaddr_protocols_err*>(arg);
+        test->one_sided_disconnect(test->sender(), UCP_EP_CLOSE_MODE_FORCE);
+        while (m_err_count == 0) {
+            test->short_progress_loop();
+        }
+    }
+
+    test_ucp_sockaddr_protocols_err() {
+        set_tl_timeouts(m_env);
+    }
+
+    void init() {
+        test_ucp_sockaddr_protocols::init();
+    }
+
+    void test_tag_send_recv(size_t size, bool is_exp,
+                            bool is_sync = false) {
+        /* warmup */
+        test_ucp_sockaddr_protocols::test_tag_send_recv(size, is_exp, is_sync);
+
+        /* run error-handling test */
+        test_ucp_sockaddr_protocols::test_tag_send_recv(size, is_exp, is_sync,
+                                                        (get_variant_value() &
+                                                         SEND_STOP) ?
+                                                        send_stop : NULL,
+                                                        (get_variant_value() &
+                                                         RECV_STOP) ?
+                                                        recv_stop : NULL,
```

Comment:
can we check the variant inside test_tag_send_recv?

Reply:
we can't since this is the function from the base class
but we need to check variants and apply them only warmup

Reply:
any other way to simplify this code?

Reply:
fixed as you suggested, but it was required to move functions to `test_ucp_sockaddr_protocols` class